lms/djangoapps/courseware/access_utils.py

```diff
@@ -92,6 +93,14 @@ def check_start_date(user, days_early_for_beta, start, course_key, display_error
         if should_grant_access:
             return ACCESS_GRANTED
 
+        # Improve performance by skipping DB read against EnterpriseCourseEnrollment if the user is anonymous.
+        if not user.is_anonymous:
+            enterprise_enrollments = EnterpriseCourseEnrollment.objects.filter(
+                course_id=course_key, enterprise_customer_user__user_id=user.id
+            )
+            if enterprise_enrollments.exists():
+                return StartDateEnterpriseLearnerError(start, display_error_to_user=display_error_to_user)
```

Comment:
[curious, sanity check] I wonder if we'll want to include a check whether the enterprise customer has the enterprise learner portal enabled here (i.e., `enable_learner_portal`). I believe there are still a handful of older, but active customers without the learner portal enabled, where redirecting to the learner portal would result in a 404 screen in the learner portal UI.

Reply:
I definitely wasn't thinking of that, seem like something worth addressing.

Reply:
Another edge case to think about: does it matter if the user has been de-linked from the EnterpriseCustomer?

Reply:
[idea] Would it make sense to try to extract the current enterprise customer from the Django session? For example, the learner dashboard MFE (frontend-app-learner-dashboard) relies on this `get_enterprise_customer` method ([source](https://github.com/openedx/edx-platform/blob/93099c384d5a1ae97083e0e6dbd0040a00cc41a4/lms/djangoapps/learner_home/views.py#L207-L217)) in its API to gather the metadata about the enterprise customer, if any.

Perhaps we could do something similar here? It predominantly relies on `enterprise_customer_from_session_or_learner_data` ([source](https://github.com/openedx/edx-platform/blob/93099c384d5a1ae97083e0e6dbd0040a00cc41a4/openedx/features/enterprise_support/api.py#L849)) which tries to extract out the active customer from the Django session (which, in theory, should be set for the learner portal customer they were viewing). If not set in the session, falls back to trying to extract a linked enterprise customer from the database.